jsonld templating playground

C-Loftus opened the pull request:
- while written guides are very useful, they don't allow users to change individual keys and experiment with templating
    - users who need to template their data for Geoconnex may need to see a dynamic example for them to fully understand the instructions
- it is helpful to have all three (raw json, the template, and the templated output) next to each other. In a standard docs page, this is too cramped since we have the sidebar and other items in the table of contents

In the page I added, users have 3 code editors. These are live editors with syntax highlighting and we can add schema validation if we want in the future as well. When they press the button, it applies the template to the raw json by calling a simple python backend service which applies the jinja template
-  If this seems like a reasonable idea, I can add pre-loaded examples so the users can select from the select box to switch been location/dataset oriented examples  

Please don't merge since this requires a live python backend that is trivial and I have implemented locally, but this is not yet deployed to google cloud

![image](https://github.com/user-attachments/assets/a7c4a2e3-a67a-43c6-9e51-3141bb280fc5)

C-Loftus commented:
Working with the location-oriented example.

![image](https://github.com/user-attachments/assets/e308033d-a30e-4967-9219-07ac5054e3b3)

ksonda commented:
This is a really good idea. I would extend it, rather than starting with some raw GeoJSON output, have more of a builder, similar to https://technicalseo.com/tools/schema-markup-generator/

C-Loftus commented:
Sounds good, thanks for that link!

I agree there is value in a structured playground with dropdown options to guide, although I also think there is value in keeping an unstructured playground as another separate page (since a user may want to experiment generally / copy paste an arbitrary template/response). 

The base functionality for the unstructured playground is close to done, just need the backend service live. Need to get a better idea of how we will  be deploying back-end infra generally as we refactor scheduler (i.e. docker swarm vs k8s vs separate individual service etc). I can deploy it more simply if we don't want to be blocked on that.

ksonda commented:
If it can be containerized and accessed via http get/post then google cloud run is fine as an interim. But you should discuss with @adplincinst

C-Loftus commented:
Ben is going to walk me through the Google cloud credentials this upcoming week and I will set up cloud run; in the meantime, I experimented with adding sparnatural to this.
The react integration for sparnatural is a bit finicky since it is written in jquery but our build system is for react. Maybe in the future we can add it or have a separate site specifically for it, but at the moment it doesn't seem to be a good fit for our docs site.

Alternative, there is a langchain LLM loader for rdf graphs which I looked at as well: https://python.langchain.com/docs/integrations/graphs/rdflib_sparql/ 
While I think LLM's are overhyped, I do think that for a playground/exploration scenario, having the ability to generate sparql queries with natural language that conform to our schema makes it really easy for people to experiment. Another stretch goal, but I thought it was cool.


![image](https://github.com/user-attachments/assets/1e323161-5ecf-456d-ba34-124292f98476)

ksonda commented:
Spanatural or similar would be good if we can load it with the relevant types for us. Which we haven't actually specified they well. Probably some schema.org types for administrative boundaries and cuahsi site types.

C-Loftus commented:
@ksonda This should be good if you want to take a look at it. Made the UI a bit nicer and now have the Google cloud infra up and running for it. 

@webb-ben let me know if you think I should change anything in the `## Templating Help` part of the templating playground. I wasn't sure if there is a better way to explain why pygeoapi has us use `data` for the `properties` key.

webb-ben commented:
> @webb-ben let me know if you think I should change anything in the `## Templating Help` part of the templating playground. I wasn't sure if there is a better way to explain why pygeoapi has us use `data` for the `properties` key.

``data`` is the python dictionary of the feature JSON-LD response that pygeoapi automatically generates. This does include flattening all of the objects in the properties block to the top level ([ref](https://github.com/geopython/pygeoapi/blob/master/pygeoapi/linked_data.py#L211)). Not sure the cleanest way to explain that. But for sure the [example JSON-LD ](https://api.iodh.internetofwater.app/collections/usgs-sta/items/'AR008-331856091114601'?f=jsonld) should not have properties block

C-Loftus commented:
> But for sure the [example JSON-LD ](https://api.iodh.internetofwater.app/collections/usgs-sta/items/'AR008-331856091114601'?f=jsonld)should not have properties block

I think I am still a bit confused with this. If it  `does include flattening all of the objects in the properties block to the top level` wouldn't we want it in the example? Or are you saying it does it implicitly? 

In this PR for sta pygeoapi: https://github.com/cgs-earth/sta-pygeoapi/pull/5,  I had added [tests](https://github.com/cgs-earth/sta-pygeoapi/blob/main/tests/runner.py#L15)  so just confirming then, even though the templates were right, my understanding was incorrect? But our templates still worked since I replaced properties with data the same way pygeoapi does it?

My understanding now is as follows. I will enumerate this very clearly in the docs, but would be helpful to confirm this:

- there is the default output JSON-LD document in pygeoapi for a given feature
- when we want to customize that, pygeoapi sources our template
- pygeoapi replaces the "properties" key with "data" in the original JSON(?) before the template is applied
- pygeoapi renders the JSON-LD by applying the template to the mutated JSON

webb-ben commented:
Once we also have settled on the code for the function in GCP can we archive that somewhere in this repository?

ksonda commented:
The template playground doesnt' seem to work for some reason. Maybe we need to pick a simpler example, like from one of the state gage endpoints where wha'ts coming through is just one table as geojson

C-Loftus commented on src/backend/main.py:
```diff
@@ -0,0 +1,57 @@
+    """Mimics the to_json filter in pygeoapi"""
+    return json.dumps(dict_, indent=4 if pretty else None, separators=(',', ':'))
+
+@functions_framework.http
+def templating(request: flask.Request):
+    """HTTP Cloud Function.
+    Args:
```
@webb-ben here is the backend code

C-Loftus commented on src/components/Playground.tsx:
```diff
@@ -0,0 +1,345 @@
+
+    try {
+      const dataToSend = JSON.parse(raw);
+      const templatingRequest = {
+        source_values: {
+        // we create a top level data key to mimic the fact pygeoapi does it
+          data: dataToSend,
+        },
+        template: template,
+      };
```
instead of doing this on the backend, i think it is easiest to just wrap the jsonld with a data key to mimic what pygeoapi is doing. Works for me. But wanted to confirm that this is reasonable

C-Loftus commented on docs/playground/templating.md:
```diff
@@ -0,0 +1,43 @@
+Notes on templating within pygeoapi
+</div>
+
+When pygeoapi applies a Jinja template, it considers the entire source JSON-LD to be the top-level key named `data`. Otherwise, the Jinja templating in pygeoapi is standard and follows expected patterns.
```
Correct?

C-Loftus reviewed (commented):
@ksonda if it wasn't working that is probably since I was working on making some changes after Ben and I aligned this morning. Should be working now.

Ben, please feel free to suggest / edit directly anything if I am still misunderstanding.
